kevlar/include/kevlar_bits/model/binomial/berry_inla.hpp

```diff
@@ -0,0 +1,369 @@
+            double det = 1;
+            for (int i = 0; i < ARMS; i++) {
+                for (int k = i + 1; k < ARMS; k++) {
+                    c = hess_inv[k][i] / hess_inv[i][i];
+                    for (int j = i; j < ARMS; j++) {
+                        hess_inv[k][j] = hess_inv[k][j] - c * hess_inv[i][j];
+                    }
+                }
+            }
+            for (int i = 0; i < ARMS; i++) {
+                det *= hess_inv[i][i];
+            }
+
+            // calculate p(sigma^2 | y) along with its normalization
+            auto sigma2_post = std::exp(logjoint + 0.5 * std::log(det));
+            sigma2_integral += sigma2_post * quad_wts[s];
+
+            // calculate exceedance probabilities, integrated over sigma2
+            for (int i = 0; i < ARMS; i++) {
+                double normalized =
+                    (efficacy_thresholds[i] - t[i]) / theta_sigma[i];
+                double exc_sigma2 = 0.5 * (erf(-normalized * M_SQRT1_2) + 1);
+                exceedances[i] += exc_sigma2 * sigma2_post * quad_wts[s];
+            }
+        }
+
+        // normalize the sigma2_post distribution
+        double inv_sigma2_integral = 1.0 / sigma2_integral;
+        for (int i = 0; i < ARMS; i++) {
+            exceedances[i] *= inv_sigma2_integral;
+        }
+        return exceedances;
+    }
+};
+
+template <class ValueType, int ARMS>
+template <class _GenType, class _ValueType, class _UIntType,
+          class _GridRangeType>
+struct BerryINLA<ValueType, ARMS>::SimGlobalState
+    : SimGlobalStateFixedNDefault<_GenType, _ValueType, _UIntType,
+                                  _GridRangeType> {
+    struct SimState;
+
+    using base_t = SimGlobalStateFixedNDefault<_GenType, _ValueType, _UIntType,
+                                               _GridRangeType>;
+    using typename base_t::gen_t;
+    using typename base_t::grid_range_t;
+    using typename base_t::interface_t;
+    using typename base_t::uint_t;
+    using typename base_t::value_t;
+
+    using sim_state_t = SimState;
+
+   private:
+    using model_t = BerryINLA;
+    const model_t& model;
+
+   public:
+    SimGlobalState(const model_t& model, const grid_range_t& grid_range)
+        : base_t(model.n_arm_samples, grid_range), model(model) {}
+
+    std::unique_ptr<typename interface_t::sim_state_t> make_sim_state(
+        size_t seed) const override {
+        return std::make_unique<sim_state_t>(*this, seed);
+    }
+};
+
+template <class ValueType, int ARMS>
+template <class _GenType, class _ValueType, class _UIntType,
+          class _GridRangeType>
+struct BerryINLA<ValueType, ARMS>::SimGlobalState<
+    _GenType, _ValueType, _UIntType, _GridRangeType>::SimState
+    : base_t::sim_state_t {
+   private:
+    using outer_t = SimGlobalState;
+
+   public:
+    using base_t = typename outer_t::base_t::sim_state_t;
+    using typename base_t::interface_t;
+
+   private:
+    const outer_t& outer_;
+    std::mt19937 gen;
```

Comment:
Lol you generalized `GenType` and hard-coded `std::mt19937` xD. You can remove this actually - a generator is already allocated in the base class.

Reply:
Oh yeah, I just copied over from DirectBayes and was sort of on a beeline to get this done and stop spending time on it! I can fix that and yeah I just saw the change with the base class from your PR. Thanks.

Reply:
Dope